### PR TITLE
Port node-cpp-skel updates

### PR DESCRIPTION
**Pull request by GretaCB**
Per https://github.com/mapbox/node-fontnik/issues/142

### Next Actions
- [x] support for more node versions (https://github.com/mapbox/node-fontnik/issues/138)
- [x] moving the secure param's to the travis settings UI (no longer storing in .travis.yml) per https://github.com/mapbox/node-cpp-skel/blob/master/docs/publishing-binaries.md
- [x] ensuring the code is correctly formatted with clang-format and all bugs found by clang-tidy are fixed
- [ ] ~~mason-js is used (once that lands in node-cpp-skel)~~
- [ ] Add modeul_utils.hpp for CallbackError method (?) and integrate into error handling
- [ ] Update docs
- [ ] Update benchmarks (?)
- [x] Update `.gitignore`
- [x] Update binding.gyp to point to new instal script to replace `install_mason.sh`
- [x] Fix new flagged compile errors
- [x] Resolve V8/Node specific compile warnings for code that we dont control
- [x] Remove unneeded diagnostics per https://github.com/mapnik/mapnik/issues/3800

@mapbox/core-tech

### CODE REVIEW

**Comment by GretaCB** [comment]
Initial port committed. Next steps are solving the new compile errors due to added flags and getting Travis running properly.

File: scripts/install_deps.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  mason install $1 $2
+  mason link $1 $2
+}
+
+# setup mason
+./scripts/setup.sh --config local.env
+source local.env
+
+install boost 1.63.0
+#link boost 1.63.0 # is linking needed?
```
Comment (springmeyer) [inline scripts/install_deps.sh]: 👍  that can be removed: linking is done on line 8 above.

**Comment by springmeyer** [comment]
This is looking fantastic @GretaCB - nice work. 👍  to merging once travis is green. All other items I think it is okay to hit as followups.

**Comment by springmeyer** [comment]
Took a look at remaining compiler warnings/errors to help @GretaCB when you pick this back up:

```
../src/glyphs.cpp:104:66: warning: use of old-style cast [-Wold-style-cast]
    uv_queue_work(uv_default_loop(), &baton->request, LoadAsync, (uv_after_work_cb)AfterLoad);
```

This one is because we are using the old C style cast. This is harmless, but bad practice in C++ where we want to be explicit. The question here is what C++ cast to use: `static_cast`? `reinterpret_cast`? I did not know so I tried `static_cast` and got a compiler error - nope. So then I cheated and looked inside `nan.h` to see what its wrapper around `ub_queue_work` does. I don't think we need to port to using Nan::AsyncQueueWorker as part of this effort, btw. But we can learn from it: https://github.com/nodejs/nan/blob/b2bb63d68b8ae623a526b542764e1ac82319cb2c/nan.h#L1870-L1877. So, `reinterpret_cast` will fix it like:

```diff
diff --git a/src/glyphs.cpp b/src/glyphs.cpp
index 26fca2d..e51203d 100644
--- a/src/glyphs.cpp
+++ b/src/glyphs.cpp
@@ -101,7 +101,7 @@ NAN_METHOD(Load) {
     }
 
     LoadBaton* baton = new LoadBaton(obj,info[1]);
-    uv_queue_work(uv_default_loop(), &baton->request, LoadAsync, (uv_after_work_cb)AfterLoad);
+    uv_queue_work(uv_default_loop(), &baton->request, LoadAsync, reinterpret_cast<uv_after_work_cb>(AfterLoad));
 }
 
 NAN_METHOD(Range) {
@@ -143,7 +143,7 @@ NAN_METHOD(Range) {
                                        info[1],
                                        start->IntegerValue(),
                                        end->IntegerValue());
-    uv_queue_work(uv_default_loop(), &baton->request, RangeAsync, (uv_after_work_cb)AfterRange);
+    uv_queue_work(uv_default_loop(), &baton->request, RangeAsync, reinterpret_cast<uv_after_work_cb>(AfterRange));
 }
 
 struct ft_library_guard {
```

--------

```
../src/glyphs.cpp:144:47: warning: implicit conversion loses integer precision: 'int64_t' (aka 'long long') to 'std::uint32_t' (aka 'unsigned int')
      [-Wshorten-64-to-32]
                                       start->IntegerValue(),
                                       ~~~~~~~^~~~~~~~~~~~~~
```

This one feels important to carefully fix, including an audit of whether we protect from overflows and edge cases. Getting ride of the warning is likely as easy as a `static_cast` to convert the integers, but do we protect against overflows? Do we test overflows? Do we have test coverage to show we test these edge cases? Thank goodness, looks like we do: https://codecov.io/gh/mapbox/node-fontnik/src/533b2a85363e0c20addc3034c5010cbe8695be8c/src/glyphs.cpp#L127...131, so I think we can just add `static_cast` here and move along.

```
../src/glyphs.cpp:198:38: warning: implicit conversion loses integer precision: 'FT_Long' (aka 'long') to 'int' [-Wshorten-64-to-32]
                num_faces = ft_face->num_faces;
                          ~ ~~~~~~~~~^~~~~~~~~
```

--------


This one is minor: I really doubt that this is going to ever be a problem: usually fonts with more than one face only have a handful. It seems silly to need a `long` type to represent faces (since I doubt a font would have so many faces to overflow `int`). But nevertheless we should fix this. Looks reasonable to do this:

```
@@ -185,7 +185,7 @@ void LoadAsync(uv_work_t* req) {
             /* LCOV_EXCL_END */
         }
         FT_Face ft_face = 0;
-        int num_faces = 0;
+        FT_Long num_faces = 0;
         for ( int i = 0; ft_face == 0 || i < num_faces; ++i )
         {
             ft_face_guard face_guard(&ft_face);
```


--------

```
../src/glyphs.cpp:307:44: warning: use of old-style cast [-Wold-style-cast]
                FT_Set_Char_Size(ft_face,0,(FT_F26Dot6)(size * (1<<6)),0,0);
                                           ^           ~~~~~~~~~~~~~~~
```

Another dodgy C cast that we should convert to C++ style. Another case where the question is which cast. I usually just check if `static_cast` works and if it compiles, then I'm happy. In this case I also remembered that this code was originally ported from mapnik, and I did a pass to clean up C style casts in mapnik a while ago. https://github.com/mapnik/mapnik/blob/master/src/text/face.cpp#L56 shows that we are doing `static_cast` there now.

-------

```
../src/glyphs.cpp:327:54: warning: implicit conversion turns floating-point number into integer: 'double' to '::google::protobuf::int32' (aka 'int')
      [-Wfloat-conversion]
                    mutable_glyph->set_top(glyph.top - glyph.ascender);
                                   ~~~~~~~ ~~~~~~~~~~^~~~~~~~~~~~~~~~
```

This one is hard. And I don't know what to do, so I'll just drop my brain dump/as I think text....

What type does `set_top` take? What type is `glyph.top` and what type is `glyph.ascender`? The compiler seems to be telling us that `glyph.top` is a `double` and we are converting it to an `int` which sounds bad: we should not be letting a conversion like this happen implicitly - that smells like trouble and potential bugs. It should be explicit and controlled. But what is the `int`: is it `glyph.ascender` or is the `int` the `set_top` argument? Based on the `top` type at https://github.com/mapbox/node-fontnik/blob/f327e89be3a4611090505211d336b46b732cda13/proto/glyphs.proto#L20 I think that `set_top` is the place we are asking for an int and implicit conversion is happening... but then were the heck is a double coming from? It looks like the `glyph` variable is a `sdf_glyph_foundry::glyph_info`. Hrm and that lives over at https://github.com/mapbox/sdf-glyph-foundry. As, so the `double` is coming from https://github.com/mapbox/sdf-glyph-foundry/blob/master/include/mapbox/glyph_foundry.hpp#L52-L54. but why is that a double: does it need to be? Can we safely convert to an int? When should we? Ugh. Hmm, going deeper it looks like `ascender` is actually an `int` at least until https://github.com/mapbox/sdf-glyph-foundry/blob/6b1632f82e9b444c586ba44fcaba876ebfb1fce1/include/mapbox/glyph_foundry_impl.hpp#L225. Should it be a double? ugh. Looks like it is pulling info from the FT_Face `ascender` and that is ultimately an `FT_Short` (aka `signed short`). So, I guess that means it is our mistake that we are ending up using `double` at all since we are actually working in integers. Unless we should actually doing division in doubles in `sdk-glyph-foundary` to avoid a potential bug from integer division? Is there a potential bug if the division is happening in integer space. Feels like it. I see the code is `int ascender = ft_face->size->metrics.ascender / 64;`. I think the `metrics.ascender` is an `FT_Pos` per https://www.freetype.org/freetype2/docs/reference/ft2-base_interface.html#FT_Size_Metrics which is an integer type. Certainly is if the  Wrote a test of it at signed long  and yes, looks like this code is probably buggy. So, if I'm seeing this right, my take home is that we should not fix this warning but rather fix `sdf-glyph-foundry`

**Comment by GretaCB** [comment]
Thanks so much @springmeyer . I was able to fix the majority of them, except for the last doozy you walked through just above, another related to the doozy, and a final one related to [`baton->message.size()`](https://github.com/mapbox/node-fontnik/blob/f327e89be3a4611090505211d336b46b732cda13/src/glyphs.cpp#L355)

Also running the tests resulted in a seg fault. Going to double check the README to make sure I have it setup correctly locally.

**Comment by GretaCB** [comment]
yeah, looks like master tests are 👍 , but the seg fault is related to this branch. Going to dig a bit and see if the casts are causing trouble.

**Comment by springmeyer** [comment]
@GretaCB great progress. I just ran your latest code locally and am not seeing a segfault, but I am seeing:

```
FATAL ERROR: invalid array length Allocation failed - process out of memory
```

I suspect that one of the casts is causing trouble. While casts will avoid a warning, casts, if done incorrectly may produce the wrong result. Maybe a small number is becoming large, or a signed number is overflowing when converted to unsigned?

**Comment by springmeyer** [comment]
Per voice with @GretaCB - I took a look at the crashes. My methods were:

 - confirm they still happen with latest code locally. they do.
 - wondered: when did they start?
 - Did `git revert d2db592` (the last commit) and rebuild with `WERROR=false make && make test` and noticed they were gone and all tests pass again
 - went back to latest code by undoing the revert with `git reset HEAD^ && git checkout .`. I have a bash alias that does this with `alias git-uncommit="git reset HEAD^ && git checkout ."`
 - confirmed the crash was back with  `WERROR=false make && make test`.
 - Ran `logbt -- make test` to see where the crash is happening and generated a backtrace.
 - Looked through the backtrace until `node_fontnik` showed up in a thread callstack:

```
    frame #9: 0x00000001001631ac node`v8::Array::New(v8::Isolate*, int) + 108
    frame #10: 0x00000001024865f8 fontnik.node`node_fontnik::AfterLoad(uv_work_s*) + 136
```
 - Felt stumped. Looked back at https://github.com/mapbox/node-fontnik/pull/144/commits/d2db592181483df32ec2e82c90dac706577d4185#diff-43b5196f9bcadc56ba40265a6712ab38R146, since we know this commit introduced the problem, just unclear why. What changed is the `reinterpre_cast` instead of an old "C style" cast - that should be fine and should not create a crash.
 - Did not see anything wrong still. Then looked a second time and saw it by squinting! Copy/paste error: `(uv_after_work_cb)AfterRange` became `reinterpret_cast<uv_after_work_cb>(AfterLoad)` and should instead be `reinterpret_cast<uv_after_work_cb>(AfterRange)`. This is facinating and so hard to see!
 - I presume what is happening is that we are crashing at https://github.com/mapbox/node-fontnik/blob/d2db592181483df32ec2e82c90dac706577d4185/src/glyphs.cpp when we try to convert a RangeBaton into a LoadBaton. This is dangerous because of how all node addon's are forced to cast like this to pass C++ structures from C/libuv and back. This would not be possible if we were able to pass a reference of the LoadBaton rather than a pointer - this mistake would ideally create a compiler error...
 - I wonder if we could add some smart checking with http://en.cppreference.com/w/cpp/language/static_assert that the type is what we would expect so that we could protect from a problem like this in the future?

**Comment by GretaCB** [comment]
Bless you, @springmeyer 🙇‍♀️
